### PR TITLE
Export above watermark

### PR DESCRIPTION
**Pull request by rain-on**

### CODE REVIEW

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java [hunk 1]
```diff
@@ -102,58 +107,101 @@ private void populateInterchangeData(final JsonGenerator jsonGenerator) {
...
+    if (watermark.getSlot() == null) {
+      LOG.warn(
+          "No block slot low watermark exists for {}, producing empty block listing",
+          validator.getPublicKey());
+    } else {
+
+      signedBlocksDao
+          .findAllBlockSignedBy(handle, validator.getId())
+          .forEach(
+              b -> {
```
Comment (usmansaleem) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:142]: nit .. `b` signifies `block`??

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java [hunk 2]
```diff
@@ -102,58 +107,101 @@ private void populateInterchangeData(final JsonGenerator jsonGenerator) {
...
+      final Validator validator,
+      final JsonGenerator jsonGenerator)
       throws IOException {
     jsonGenerator.writeArrayFieldStart("signed_attestations");
 
+    if (watermark.getSourceEpoch() == null || watermark.getTargetEpoch() == null) {
+      LOG.warn(
+          "Missing attestation low watermark for {}, producing empty attestation listing",
+          validator.getPublicKey());
+    }
```
Comment (usmansaleem) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:178]: is that just a warning or do we want to return early?

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java [hunk 3]
```diff
@@ -102,58 +107,101 @@ private void populateInterchangeData(final JsonGenerator jsonGenerator) {
...
-                      new tech.pegasys.web3signer.slashingprotection.interchange.model
-                          .SignedAttestation(
-                          a.getSourceEpoch(), a.getTargetEpoch(), a.getSigningRoot().orElse(null));
-              try {
-                mapper.writeValue(jsonGenerator, jsonAttestation);
-              } catch (final IOException e) {
-                throw new UncheckedIOException(
-                    "Failed to construct a signed_attestations entry in json", e);
+              if ((a.getSourceEpoch().compareTo(watermark.getSourceEpoch()) >= 0)
+                  && (a.getTargetEpoch().compareTo(watermark.getTargetEpoch()) >= 0)) {
```
Comment (usmansaleem) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:185]: does compareTo works against `null` values? (related to previous comment).

**Review by usmansaleem — APPROVED** [review]
LGTM - minor concerns around compareTo with potentially null values.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java [hunk 4]
```diff
@@ -102,58 +107,101 @@ private void populateInterchangeData(final JsonGenerator jsonGenerator) {
...
+      final Handle handle, final Validator validator, final JsonGenerator jsonGenerator)
       throws IOException {
     jsonGenerator.writeStartObject();
     jsonGenerator.writeStringField("pubkey", validator.getPublicKey().toHexString());
-    writeBlocks(h, validator, jsonGenerator);
-    writeAttestations(h, validator, jsonGenerator);
+    final Optional<SigningWatermark> watermark =
+        lowWatermarkDao.findLowWatermarkForValidator(handle, validator.getId());
+    if (watermark.isEmpty()) {
+      LOG.warn("No low watermark available, producing empty export.");
```
Comment (jframe) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:117]: In this case we probably want to exit and not produce an export at all

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java [hunk 5]
```diff
@@ -102,58 +107,101 @@ private void populateInterchangeData(final JsonGenerator jsonGenerator) {
...
-                          b.getSlot(), b.getSigningRoot().orElse(null));
-              try {
-                mapper.writeValue(jsonGenerator, jsonBlock);
-              } catch (final IOException e) {
-                throw new UncheckedIOException(
-                    "Failed to construct a signed_blocks entry in json", e);
-              }
-            });
+    if (watermark.getSlot() == null) {
+      LOG.warn(
```
Comment (jframe) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:134]: Think this is the same as not having a signing record where we don't create a signing record. Perhaps the top level check need to verify that we have all non-null values for the watermark record.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java (hunk 3 above)
Comment (jframe) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:185]: Should this just be target > 0? The attestation validator rule for hasTargetOlderThanWatermark excludes signing anything < watermark excluding the equals case. I believe this difference between source and target was from the interchange spec rules.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java [hunk 6]
```diff
@@ -102,58 +107,101 @@ private void populateInterchangeData(final JsonGenerator jsonGenerator) {
...
+                            a.getSigningRoot().orElse(null));
+                try {
+                  mapper.writeValue(jsonGenerator, jsonAttestation);
+                } catch (final IOException e) {
+                  throw new UncheckedIOException(
+                      "Failed to construct a signed_attestations entry in json", e);
+                }
+              } else {
+                LOG.debug(
+                    "Ignoring attestation with source epoch {}, and taget epoch {}, for validator {}, as is below watermark",
```
Comment (jframe) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:201]: typo /taget/target

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java (hunk 1 above)
Comment (rain-on) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:142]: updated

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java (hunk 2 above)
Comment (rain-on) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:178]: early exit

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java (hunk 3 above)
Comment (rain-on) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:185]: so the watermarks cannot be null anymore - so does that solve the issue?

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2SubCommand.java [hunk 7]
```diff
@@ -42,7 +42,10 @@
 
   @Spec CommandSpec spec;
 
-  @Command(name = "export", description = "Export slashing protection db to json file")
+  @Command(
```
[the remark this answers is not in the document]
Comment (rain-on) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2SubCommand.java]: reworked extensively in another PR.

File: slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/InterchangeExportIntegrationTest.java [hunk 8]
```diff
@@ -56,36 +48,23 @@ void canCreateDatabaseWithEntries() throws IOException {
...
-          h -> {
-            for (int b = 0; b < TOTAL_BLOCKS_SIGNED; b++) {
-              signedBlocks.insertBlockProposal(
-                  h, new SignedBlock(validatorId, UInt64.valueOf(b), Bytes.fromHexString("0x01")));
-            }
-          });
+      slashingProtection.registerValidators(List.of(validatorPublicKey));
+
+      for (int b = 0; b < TOTAL_BLOCKS_SIGNED; b++) {
+        insertBlockAt(UInt64.valueOf(b), validatorPublicKey);
```
[the remark this answers is not in the document]
Comment (rain-on) [reply slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/InterchangeExportIntegrationTest.java:54]: Done.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java (hunk 6 above)
Comment (rain-on) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Exporter.java:201]: done

**Review by jframe — APPROVED** [review]

**Pull request merged**